x/expiration/types/msgs.go

```diff
@@ -0,0 +1,161 @@
+package types
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	metadatatypes "github.com/provenance-io/provenance/x/metadata/types"
+)
+
+// Compile time interface checks
+var (
+	_ sdk.Msg = &MsgAddExpirationRequest{}
+	_ sdk.Msg = &MsgExtendExpirationRequest{}
+	_ sdk.Msg = &MsgInvokeExpirationRequest{}
+)
+
+// MustAccAddressFromBech32 converts a Bech32 address to sdk.AccAddress
+// Panics on error
+func MustAccAddressFromBech32(s string) sdk.AccAddress {
+	accAddress, err := sdk.AccAddressFromBech32(s)
+	if err != nil {
+		panic(err)
+	}
+	return accAddress
+}
+
+// private method to convert an array of strings into an array of Acc Addresses.
+func stringsToAccAddresses(signers []string) []sdk.AccAddress {
+	addresses := make([]sdk.AccAddress, len(signers))
+	for i, address := range signers {
+		addresses[i] = MustAccAddressFromBech32(address)
+	}
+	return addresses
+}
+
+// private method that does a simple validation check that
+// doesn't require access to any other information.
+func validateBasic(expiration Expiration, signers []string) error {
+	if err := expiration.ValidateBasic(); err != nil {
+		return err
+	}
+	if len(signers) == 0 {
+		return ErrMissingSigners
+	}
+	return nil
+}
+
+// private method that checks an address string is bech32 or MetadataAddress type
+func validateAddress(s string) error {
+	if _, err := sdk.AccAddressFromBech32(s); err != nil {
+		// check if we're dealing with a MetadataAddress
+		if _, err2 := metadatatypes.MetadataAddressFromBech32(s); err2 != nil {
+			return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress,
+				fmt.Sprintf("invalid module asset id: %s", err.Error()))
+		}
```

Comment:
Again, `sdk.AccAddressFromBech32(s)` will return an error if given a metadata address.

Reply:
It first checks to see if it is an AccAddress and when that fails it checks to see if we are dealing with a metadata address.

Reply:
Ah. I see it now. It can be either a regular chain address or a metadata address.

This presents another problem though. We don't want the expiration module to import anything from the metadata module. As such, we probably can't have this check in the `types` package since there will need to be some sort of function registry that the expiration module can use to check if an address is usable. The interface would be something like `IsExpirableAddress(string) bool`. And the expiration module's `validateAddress` function would then loop through those, returning false only if all of them return false (maybe also have the `sdk.AccAddressFromBech32(s)` check, but maybe not). Or maybe it'd be more like an antehandler setup where you chain them together, each one calling the next. I'm not sure which is best.

For now though, maybe just use the `bech32.DecodeAndConvert` function from `"github.com/cosmos/cosmos-sdk/types/bech32"` library directly, e,.g. `hrp, bz, err := bech32.DecodeAndConvert(address)`.